tiatoolbox/wsicore/wsimeta.py

```diff
@@ -173,7 +176,116 @@ def validate(self):
         if all(x is None for x in [self.objective_power, self.mpp]):
             warnings.warn("Unknown scale (no objective_power or mpp)")
 
-        return passed
+        return passed  # noqa
+
+    def level_downsample(
+        self,
+        level: Union[int, float],
+    ) -> float:
+        """Get the downsample factor for a level.
+
+        For non-integer values of `level`, the downsample factor is
+        linearly interpolated between from the downsample factors of
+        the level below and the level above.
+
+        Args:
+            level (int or float): Level to get downsample factor for.
+
+        Returns:
+            float: Downsample factor for the given level.
+
+        """
+        level_downsamples = self.level_downsamples
+        if isinstance(level, int) or int(level) == level:
+            # Return the downsample for the level
+            return level_downsamples[int(level)]
+        # Linearly interpolate between levels
+        floor = int(np.floor(level))
+        ceil = int(np.ceil(level))
+        floor_downsample = level_downsamples[floor]
+        ceil_downsample = level_downsamples[ceil]
+        return np.interp(level, [floor, ceil], [floor_downsample, ceil_downsample])
+
+    def relative_level_scales(
+        self, resolution: Resolution, units: str
+    ) -> List[np.ndarray]:
+        """Calculate scale of each level in the WSI relative to given resolution.
+
+        Find the relative scale of each image pyramid / resolution level
+        of the WSI relative to the given resolution and units.
+
+        Values > 1 indicate that the level has a larger scale than the
+        target and < 1 indicates that it is smaller.
+
+        Args:
+            resolution (float or tuple(float)): Scale to calculate
+                relative to units.
+            units (str): Units of the scale. Allowed values are: mpp,
+                power, level, baseline. Baseline refers to the largest
+                resolution in the WSI (level 0).
+
+        Raises:
+            ValueError: Missing MPP metadata
+            ValueError: Missing objective power metadata
+            ValueError: Invalid units
+
+        Returns:
+            list: Scale for each level relative to the given scale and
+                units.
+
+        Examples:
+            >>> from tiatoolbox.wsicore.wsireader import WSIReader
+            >>> wsi = WSIReader.open(input_img="./CMU-1.ndpi")
+            >>> print(wsi._relative_level_scales(0.5, "mpp"))
+            [array([0.91282519, 0.91012514]), array([1.82565039, 1.82025028]) ...
+
+            >>> from tiatoolbox.wsicore.wsireader import WSIReader
+            >>> wsi = WSIReader.open(input_img="./CMU-1.ndpi")
+            >>> print(wsi._relative_level_scales(0.5, "baseline"))
```

Comment:
Duplicate comment as on line 239: This is incorrect since _relative_level_scales was moved from wsireader to wsimeta. Python will throw an error "AttributeError: 'OpenSlideWSIReader' object has no attribute '_relative_level_scales'"